dotenv-linter/src/checks/schema_violation.rs

```diff
@@ -0,0 +1,139 @@
+                        } else {
+                            return Some(Warning::new(
+                                line.number,
+                                self.name(),
+                                format!("The {} key is not a valid boolean", key),
+                            ));
+                        }
+                    }
+                    SchemaValueType::Integer => {
+                        if value.parse::<i32>().is_err() {
+                            return Some(Warning::new(
+                                line.number,
+                                self.name(),
+                                format!("The {} key is not an integer", key),
+                            ));
+                        }
+                    }
+                    SchemaValueType::Float => {
+                        if value.parse::<f32>().is_err() {
+                            return Some(Warning::new(
+                                line.number,
+                                self.name(),
+                                format!("The {} key is not a valid float", key),
+                            ));
+                        }
+                    }
+                    SchemaValueType::Email => {
+                        if !EmailAddress::is_valid(value) {
+                            return Some(Warning::new(
+                                line.number,
+                                self.name(),
+                                format!("The {} key is not a valid email address", key),
+                            ));
+                        }
+                    }
+                    SchemaValueType::Url => {
+                        if Url::parse(value).is_err() {
+                            return Some(Warning::new(
+                                line.number,
+                                self.name(),
+                                format!("The {} key is not a valid URL", key),
+                            ));
+                        }
+                    }
+                }
+            }
+        } else if !schema.allow_other_keys {
+            return Some(Warning::new(
+                line.number,
+                self.name(),
+                format!("The {} key is not defined in the schema", key),
+            ));
+        }
+
+        None
+    }
+
+    fn name(&self) -> LintKind {
+        LintKind::SchemaViolation
+    }
+
+    fn end(&mut self) -> Vec<Warning> {
+        let mut warnings = Vec::new();
+        if let Some(schema) = self.schema {
+            for (key, entry) in &schema.entries {
+                if entry.required && !self.seen_keys.contains(key) {
+                    warnings.push(Warning::new(
+                        self.last_line_number,
+                        self.name(),
+                        format!("The {} key is required", entry.key),
+                    ));
+                }
+            }
+        }
+        warnings
+    }
```

Comment:
```suggestion
pub(crate) fn check_required_keys(schema: &crate::schema::DotEnvSchema, keys: &[&str]) -> Vec<Warning> {
    let mut warnings = Vec::new();
    for (key, entry) in &schema.entries {
        if entry.required && !keys.contains(&key.as_str()) {
            warnings.push(Warning::new(
                keys.len() + 1,
                LintKind::SchemaViolation,
                format!("The {} key is required", entry.key),
            ));
        }
    }
    warnings
}
```
The end function is not too obvious, so I propose this solution.

Reply:
Btw, I think there is a better alternative.

- extract all required keys from schema in `SchemaViolationChecker::new` and keep as `HashSet`
- every call of `run` we remove the key from the hashset
- If at the end there are keys left in the `hashset`, we return warnings for them.

This should be more efficient than going through the keys we collected again. Also, we will not be forced to store all the keys from the file in memory.

This will still require the checker to have a callback, much like now, but instead of `end` I would suggest something like `remaining_warnings` or so. I believe this is even better, because instead of hardcode we will get expandable functionality for checkers.

Reply:
the 'end' call is intended as a generic callback for all checkers that may want to do a final pass of some kind

Removing the keys once seen will prevent the validation of duplicated entries, ie regex check... The second one may be the 'intended' one

Reply:
> Removing the keys once seen will prevent the validation of duplicated entries, ie regex check... The second one may be the 'intended' one

Didn't get it. What do you mean?

I expect to have something like `required_keys: HashSet<String>` and we remove the keys from it once seen. It shouldn't be used for anything but required keys

Required key present - we remove it from the set.  The logic is almost the same as current one with second pass through all keys.